arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java

```diff
@@ -308,7 +307,11 @@ public boolean hasNext() {
 
     @Override
     public ColumnarBatch next() {
-      return current;
+      if (hasNext()) {
+        return currentIterator.next();
```

Comment:
I think it would be sufficient to just call `currentIterator.next` here because the `NoSuchElementException` will be thrown implicitly when there's no other element

Reply:
Is it possible to initialize `currentIterator` elsewhere so that `next` doesn't need to call `hasNext`?

Reply:
I'm following the pattern in https://github.com/apache/iceberg/blob/90225d6c9413016d611e2ce5eff37db1bc1b4fc5/api/src/main/java/org/apache/iceberg/io/CloseableIterable.java#L204. 

Initializing currentIterator is not the only thing that needs to be done, we need to keep creating new currentIterator on every new data file. We can move the common logic to another method but it will be same logic as hasNext() and I don't see any real benefit in moving the code. 

Do you have something in mind that will improve the logic?

Reply:
I can't think of anything that would improve the logic.

I'm a bit hesitant about copying some of the logic from `CloseableIterator` as there are a few situations we've seen where that hangs. But in the absence of a complete understanding of that issue, I think it's fine to borrow form there as it's the same situation.

Also, thanks for letting me know that this came from `CloseableIterator`. I have plans to begin researching that soon and that helps. 👍

Reply:
I think this is good, and I agree with the call to `hasNext()`. We shouldn't rely on the underlying iterator for correct behavior (throwing `NoSuchElementException`) because iterators are actually hard to get right. Plus, this ensures that `hasNext()` doesn't advance the iterator.